### ISSUE TITLE
1.4.0 stopped serving chain certificates

### ISSUE THREAD

**Issue by hynek** [linkerd/linkerd#1926]
# Thanks for your help improving the project! #

## Filing a Linkerd issue ##

Issue Type:

- [x] Bug report
- [ ] Feature request

**What happened**:

After upgrading from 1.3.7, HAProxy's health check started failing with

    `bk-l5d-router-slow/l5d-local is DOWN, reason: Layer6 invalid response, info: "SSL handshake failure"`

and Python connections with

    [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed

Turns out, that with the same config and the same PEM files, linkerd stopped serving chain files that are part of `certPath`.

1.3.7:

```
depth=2 C = DE, ST = BB, L = Potsdam, O = Variomedia AG, CN = ca.vm.ag, emailAddress = ca@vm.ag
verify return:1
depth=1 CN = ca.vm.ag
verify return:1
depth=0 CN = *.host.vrmd.de
verify return:1
---
Certificate chain
 0 s:/CN=*.host.vrmd.de
   i:/CN=ca.vm.ag
 1 s:/CN=ca.vm.ag
   i:/C=DE/ST=BB/L=Potsdam/O=Variomedia AG/CN=ca.vm.ag/emailAddress=ca@vm.ag
```

1.4.0

```
depth=0 CN = *.host.vrmd.de
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = *.host.vrmd.de
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/CN=*.host.vrmd.de
   i:/CN=ca.vm.ag
```

It's *not* because of the order inside the file because if I flip them, the key fails to load.

I strongly presume this is related to the Finagle upgrade but I’m baffled that I’m the only person running into it.

**What you expected to happen**:

Nothing actually. I’d like it to keep working. 😇

**How to reproduce it (as minimally and precisely as possible)**:

`¯\_(ツ)_/¯`

**Anything else we need to know?**:

` ¯\_(ツ)_/¯`

**Comment by dadjeibaah** [comment]
Thanks for filing this issue @hynek! We will take a look at what's going on

**Comment by adleong** [comment]
@hynek Yikes, this is my top priority issue.  If you can provide any of the following information, it will greatly help tracking this down:

* What protocol? http or h2?
* Can you provide your Linkerd config?
* Can you provide your cert PEM file (without private key)?
  * If not, can you provide instructions for how to generate a cert with the same structure?

If you'd rather share any of the above privately instead of publicly, you can private message me on the Linkerd slack or email me alex@buoyant.io

**Comment by hynek** [comment]
I’m travelling and on my phone right now. I can send you the pem file and config later. I’m fairly confident it’s created using HashiCorp’s Vault CA capabilities.

**Comment by hynek** [comment]
Here’s our config template for now. Both admin and servers have the same problem.

``` yaml
---
admin:
  port: 9990
  ip: {{ linkerd.admin.address }}
  tls:
    certPath: /etc/ssl/certs/*.host.vrmd.de_all.crt
    keyPath: /etc/ssl/private/*.host.vrmd.de.pkcs8


namers:
  - kind: io.l5d.consul
    includeTag: false
    useHealthCheck: true
    token: XXX


routers:
  - protocol: http
    label: /http-consul

    interpreter:
      kind: io.l5d.mesh
      dst: /#/io.l5d.consul/.local/namerd
      experimental: true
      root: /default

    servers:
      - port: 4140
        ip: {{ linkerd.router.address }}
        addForwardedHeader:
          by:
            kind: ip
          for:
            kind: ip
        tls:
          certPath: /etc/ssl/certs/*.host.vrmd.de_all.crt
          keyPath: /etc/ssl/private/*.host.vrmd.de.pkcs8
        clearContext: {{ "true" if linkerd.edge else "false" }}

    client:
      kind: io.l5d.static
      configs:
        - prefix: /
          requeueBudget:
            percentCanRetry: 5.0
          tls:
            disableValidation: true

    service:
      kind: io.l5d.global
      responseClassifier:
        kind: io.l5d.http.retryableIdempotent5XX


telemetry:
  - kind: io.l5d.prometheus
    path: /admin/metrics/prometheus
    prefix: linkerd_


usage:
  enabled: false. # sorry :)
```

I double checked the logs if linkerd complains about cert loading but no dice.

I suspect the reason nobody ran into it yet is that we use an intermediate certificate to be able to revoke everything at once without compromising our root and that’s not that common? I suspect that’s a path for you to reproduce.

**Comment by hynek** [comment]
(so yeah totally just http)

**Comment by adleong** [comment]
I've been attempting to reproduce this but so far I can't even get it working on 1.3.7.  To my knowledge, we've never tested this particular configuration where the server serves a certificate chain of more than one cert.  When I try this, I run into null pointer exceptions in Netty's SSL code --- though it's possible I have misformatted or misconfigured my certs.

The workflow that we usually test for using intermediate CAs is that we have the server serve only the leaf cert and we configure the client with a CA chain that includes the intermediate and the root CA.  

So, long story short: I've very eager to see how your certs are structured so that I can reproduce this.

**Comment by hynek** [comment]
It feels weird to paste it here but I guess the point of certificates is to make them public. :)

<details>

```pem
-----BEGIN CERTIFICATE-----
MIIF/TCCA+WgAwIBAgIUL0bJFKrOuNxUo5sP8N8YqWIDg90wDQYJKoZIhvcNAQEL
BQAwEzERMA8GA1UEAxMIY2Eudm0uYWcwHhcNMTgwMjEyMTAzNDQwWhcNMjMwMjEx
MTAzNTA4WjAZMRcwFQYDVQQDEw4qLmhvc3QudnJtZC5kZTCCAiIwDQYJKoZIhvcN
AQEBBQADggIPADCCAgoCggIBALw5UQkUzY8uQoxBHo+z0ZB6iMAW8GN0bdE7tVpv
6ixByLEBs4oTa7tP0xGE+l5eGTF3kNIGVGfl6/XeEde3XwF2imhj7GqJzlK0xeQK
+cKRuhOwPKTv0pVoXb+qiOk8M8eGe9UNQlDjc/ilJFUfDUc+6PwJ3pf/PBGXFQA+
7vf31G3rn65V8zNxMDkhNQqborEIGT7PvYm6eXEVcj9MqquYL2g4T5O2566s0/lv
AS4rmz86P9k8Fk2VgZ22enZyMq3t7z0t6uu8SuVp1SqYmY9mZme5vMxNn+eXGOty
SrRJt8QSepmkK8sxR+5KxvCSL13rV0ou1nXzkseyNKAMCaqOplXyKDQtrYFM5++G
yBLg5bGem69DXychqfb23sEG9UUoFTabFB99Rek3jyMLjrlGZnaOV3eonhNXI0n2
deIE8683kcuKI+otq94yScfU0FhZU6QhO5qk2EKEnCNHMqxgWxWe+FfeKtkZziXv
ni42Z7sk1WlZn9ftRd+jS8cAENFNBu8uEPL1FsDEVdwXG2BFyZKKtZAoQ8qi5j+A
VBXLugKF1kuPUK0ECshJV2RYsGqcP4gjsIZqlP3t7m44lFkdO+RcnrlkbJesOC6j
hgCrjKCrkghmwBOZc+tixjKs4CWVZlrxkUv5tC0MLGJCVQ1jfMTteI+hvITQikYc
3vaZAgMBAAGjggFBMIIBPTAOBgNVHQ8BAf8EBAMCA6gwHQYDVR0lBBYwFAYIKwYB
BQUHAwEGCCsGAQUFBwMCMB0GA1UdDgQWBBR/CLt50xJk5O3JEvVzN5Fj86mfRTAf
BgNVHSMEGDAWgBRkM9ADKxNNhqd1bwjNmGRkxqZSzTBFBggrBgEFBQcBAQQ5MDcw
NQYIKwYBBQUHMAKGKWh0dHBzOi8vY2Eudm0uYWc6ODIwMC92MS92bWNhX2ludG1f
cGtpL2NhMEgGA1UdEQRBMD+CDiouaG9zdC52cm1kLmRlgglsb2NhbGhvc3SHBH8A
AAGHBH8AAQGHBAoDADuHBAoDADqHBAoDAEGHBAoGtPowOwYDVR0fBDQwMjAwoC6g
LIYqaHR0cHM6Ly9jYS52bS5hZzo4MjAwL3YxL3ZtY2FfaW50bV9wa2kvY3JsMA0G
CSqGSIb3DQEBCwUAA4ICAQB778E1dFsyLrRQ7UBsONo5f6ESsnf2GHpkL/8LEwaI
QO2c1wbaD/rUh5rubHQYhvOTRMZXtZSnCF862AL14JB2cSCRiCEW/CKCkgw9Ez8h
fVutP/2e89FZPkvubGr1ZDRfWbbi8c0VPdT1WYOIfA3dNWz7QUnIgKQZq9lBlDDV
P3iwH9MgCy/vwXaNOQesHcl/TZP4FykUQ3h89rX2/37Ie8+YKAj58NIyu9wAE7vt
/CFgMYRAYbbiiQMhRxYX+VGCYXTwVfCqU5ZO3YcK+7xjRwSfIqnSpkm+6aokGFrv
lV3bvRgaGTARSf4bteFj5/Saoyoh+P+qfq04CfHOW+6IMogdFG1j1KKnNq3grwiG
Ufs32luA9wxYkbPeYP+FyEcFnKNuwDW+/e/nTMPZ7NF4zwxo7zxrF+t0c5sNiJRC
h2+51iOzE7JHIfv7QPgybbIzTI/Dkv+ppKRCk/FYm4odcDu6JjOREQPfrzZm+U0H
8J0+qC7qW0VFfoGjknSfzUH0uYt3WZsHSQkvbz/5ppMrdxF5ET3MMsUsKgY8rJHn
6Il7MvjvOoRyIU4QVWgsLd55YliyFpoDfpEaHmnTjCOEl4OWAJ45CmLyt7ncJyqm
1/sfSiOPTRZTOJtwu7XwLnflSmrZ3BPZOVMt8DkKQI64AZDd8iVgHiQJ0pKTgFNe
TA==
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
MIIGDzCCA/egAwIBAgIUWT3/O0scNNMVUmjb5pVcTedPZj4wDQYJKoZIhvcNAQEL
BQAwcDELMAkGA1UEBhMCREUxCzAJBgNVBAgMAkJCMRAwDgYDVQQHDAdQb3RzZGFt
MRYwFAYDVQQKDA1WYXJpb21lZGlhIEFHMREwDwYDVQQDDAhjYS52bS5hZzEXMBUG
CSqGSIb3DQEJARYIY2FAdm0uYWcwHhcNMTcxMTA3MTYyOTU2WhcNMjcxMTA1MTYz
MDI2WjATMREwDwYDVQQDEwhjYS52bS5hZzCCAiIwDQYJKoZIhvcNAQEBBQADggIP
ADCCAgoCggIBAOh2J0Suj2iFAzMkUShfeo4K5UjmnLfWMctUxgPvpcPyQTuDLR67
6PsDGJMrIjU2pTZ9E7qrZQTSpt+klPa3e52qQFAzAX28SajdbJXmKukLEXDBBHP9
uH/X2Pyd5X5YN/Q5yeOMpESn/iNtztU7ZfT7cO6GI8LhDG70mdvalKaHSI6u40+p
pDAruorToa/EQxwc6vmI0fCJykL4XH2oFisCCPVivN2Ta2AhDcahAinXKK6OvXIB
mHsoxDm3und4/vQdGaS6QBk4CH1nvy/HWFKb11uqBq923F9v+L9a+CvG3ejnoDqJ
/UIiDwdoZiPWyiw32DlrZt+BC8KL2gHA5PdHjWAr55tdD8MSND/Ydo8shxphiC4R
XlJAoju3V+/UbCVYuu71/j1p5d8BplKtgNTV+lnHgCTNMZCyDP0ZENshmQx9FPtB
hk+dXMo6lQofbeURF/cs1m6uGlyWspmVzHrnH2+roekmIPOn1nRNwhwfMQNv0MXU
3vTJoBXocv5uS+i4yR1ssKvkc/szo95SjUJpMIJANUu8DetKHbIvHhCnvSJTWKL8
K/EZP6yeQkeqFegWuc9mq+Ijk+62TlFbAyGTX3UlwGCwlnWWHMr+Yc6TtPCiixmn
uTPnL9yddaZaCw6ONJkV2qm6l9Ufdg1K1ekvbt+IYKEWrehwDdRT9QbZAgMBAAGj
gf0wgfowDgYDVR0PAQH/BAQDAgEGMA8GA1UdEwEB/wQFMAMBAf8wHQYDVR0OBBYE
FGQz0AMrE02Gp3VvCM2YZGTGplLNMB8GA1UdIwQYMBaAFKcjquPJehF25pZq6BN7
2Tgzp0X8MEUGCCsGAQUFBwEBBDkwNzA1BggrBgEFBQcwAoYpaHR0cHM6Ly9jYS52
bS5hZzo4MjAwL3YxL3ZtY2Ffcm9vdF9wa2kvY2EwEwYDVR0RBAwwCoIIY2Eudm0u
YWcwOwYDVR0fBDQwMjAwoC6gLIYqaHR0cHM6Ly9jYS52bS5hZzo4MjAwL3YxL3Zt
Y2Ffcm9vdF9wa2kvY3JsMA0GCSqGSIb3DQEBCwUAA4ICAQDXUmSW34aZExbENe2Z
OwWntDWgxrjFEf1nytwR/kAnoPanrRPBJtkrwop9FmN1860wfFv1jkIQcoxEEAX2
zvJnbvWLvBxwEnkNa1mg/UZhK9ViCVt4CARloJunZ7TTxKJw2gwA/ku1u59bOpD4
9FT+ymLJQIHd4m4jWiAmX5deuFA5pxiz8ku5wOQNA2XMUmZCozB2qISI919/yben
A2gKoBDPHlHtE33X+ass1LvDw3OKzKqqL1csYUXvJ1RKbDgvc1fGdENogGnfP+wb
gdPITkUBYZlyeumuQjsCdShbuBWxAheNp0hR9r4FeYFjjrH94J0ThdLkjQfFoknF
cOvk+5/GqhYkK9bIcS9wKW9edUjxX3AEYHg0XmkusdVH0HH3YyCJNrEoHpM1TbO3
p+65S5Cb1fsyCS4dUfd+1pJuAXKO4YzC2G24FLzdoOXNrt0H8uVWA1Z2ZgUIX4zl
oIQqgPEAEUk7eRQT4ajE+4Oo4j2Rquma0Uuty32+Ur6//NWRWAnWLWA+6sxL2YGf
hyCaEhN4PDIfFeXIpdglv4xv5jGo4oA/X+3LQpUD2Yay9YB8LKouFRgseXAD9lA7
lQAAtwgX/SeULOT632qn5dqTtEeHPIgF86CNwDHOBEFxBLTF+cgPklOYdVkaLPm2
xjNQPNcW5LR26j1qdTR9hucnDA==
-----END CERTIFICATE-----
```

</details>

***

So I have two theories for you:

1. maybe even your 1.3.7 accidentally uses a newer Finagle by accident?
2. My client configuration says `disableValidation: true` because I just couldn’t find any good docs on how to get it working (my `commonName` is predictable but changing…and I couldn’t piece myself the information together from k8n-heavy blog posts…). Anyhow, it says

   >  Setting `disableValidation: true` will force the use of the JDK SSL provider which does not support client auth.

   Could that be related and the reasons why you can’t reproduce under 1.3.7? Different TLS engines are certainly a plausible reasons for a change in behavior…
